File: backends/spring-java/src/main/java/dev/stackverse/backend/common/Logging.java
package dev.stackverse.backend.common;

import org.slf4j.Logger;
import org.slf4j.event.Level;

public final class Logging {
    private Logging() {
    }

    /**
     * Emits a stable contract event using SLF4J key-value pairs, which Spring Boot
     * structured logging and the OTel Java agent export as structured fields.
     */
    public static void logEvent(Logger log, Level level, String event, String outcome, String message, Object... fields) {
        var builder = log.atLevel(level)
            .addKeyValue("event", event)
            .addKeyValue("outcome", outcome);
        for (int i = 0; i + 1 < fields.length; i += 2) {
            if (fields[i + 1] != null) {
                builder = builder.addKeyValue(String.valueOf(fields[i]), fields[i + 1]);
            }
        }
        builder.log(message);
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/common/Logging.kt
package dev.stackverse.backend.common

import org.slf4j.Logger
import org.slf4j.event.Level

/**
 * Emits a stable contract event (docs/LOGGING.md §5): `event`, `outcome`, and the
 * extra fields travel as SLF4J key-value pairs — structured members of the JSON
 * console output and of OTLP-exported records — while the message stays prose.
 * Null-valued fields are dropped so "when applicable" fields can be passed as-is.
 */
fun Logger.logEvent(level: Level, event: String, outcome: String, message: String, vararg fields: Pair<String, Any?>) {
    var builder = atLevel(level)
        .addKeyValue("event", event)
        .addKeyValue("outcome", outcome)
    for ((key, value) in fields) {
        if (value != null) {
            builder = builder.addKeyValue(key, value)
        }
    }
    builder.log(message)
}
